src/proof_system/prover.rs

```diff
@@ -0,0 +1,348 @@
+        self.cs.circuit_size()
+    }
+
+    /// Split `t(X)` poly into 4 degree `n` polynomials.
+    pub fn split_tx_poly(
+        &self,
+        n: usize,
+        t_x: &Polynomial,
+    ) -> (Polynomial, Polynomial, Polynomial, Polynomial) {
+        (
+            Polynomial::from_coefficients_vec(t_x[0..n].to_vec()),
+            Polynomial::from_coefficients_vec(t_x[n..2 * n].to_vec()),
+            Polynomial::from_coefficients_vec(t_x[2 * n..3 * n].to_vec()),
+            Polynomial::from_coefficients_vec(t_x[3 * n..].to_vec()),
+        )
+    }
+    /// Computes the quotient opening polynomial.
+    fn compute_quotient_opening_poly(
+        n: usize,
+        t_1_poly: &Polynomial,
+        t_2_poly: &Polynomial,
+        t_3_poly: &Polynomial,
+        t_4_poly: &Polynomial,
+        z_challenge: &Scalar,
+    ) -> Polynomial {
+        // Compute z^n , z^2n , z^3n
+        let z_n = z_challenge.pow(&[n as u64, 0, 0, 0]);
+        let z_two_n = z_challenge.pow(&[2 * n as u64, 0, 0, 0]);
+        let z_three_n = z_challenge.pow(&[3 * n as u64, 0, 0, 0]);
+
+        let a = t_1_poly;
+        let b = t_2_poly * &z_n;
+        let c = t_3_poly * &z_two_n;
+        let d = t_4_poly * &z_three_n;
+        let abc = &(a + &b) + &c;
+        &abc + &d
+    }
+    /// Convert variables to their actual witness values.
+    pub(crate) fn to_scalars(&self, vars: &[Variable]) -> Vec<Scalar> {
+        vars.par_iter().map(|var| self.cs.variables[var]).collect()
+    }
+    /// Resets the witnesses in the prover object.
+    /// This function is used when the user wants to make multiple proofs with the same circuit.
+    pub fn clear_witness(&mut self) {
+        self.cs = StandardComposer::new();
+    }
+
+    /// Clears all data in the Prover
+    /// This function is used when the user wants to use the same Prover to
+    /// make a proof regarding a different circuit.
+    pub fn clear(&mut self) {
+        self.clear_witness();
+        self.preprocessed_circuit = None;
+        self.preprocessed_transcript = Transcript::new(b"plonk");
+    }
+
+    /// Keys the transcript with additional seed information
+    /// Wrapper around transcript.append_message
+    pub fn key_transcript(&mut self, label: &'static [u8], message: &[u8]) {
+        self.preprocessed_transcript.append_message(label, message);
+    }
+
+    /// Creates a Proof that a circuit is satisfied
+    /// Note that if you intend to make multiple proofs, after calling this method, the user should then
+    /// call `clear_witness`. This is automatically done when `prove` is called
+    pub fn prove_with_preprocessed(
+        &self,
+        commit_key: &ProverKey,
+        preprocessed_circuit: &PreProcessedCircuit,
+    ) -> Proof {
+        let domain = EvaluationDomain::new(self.cs.circuit_size()).unwrap();
+
+        // Clone the transcript so we can do multiple proofs
+        let mut transcript = self.preprocessed_transcript.clone();
```

Comment:
This should be modified via the circuit passed in

Reply:
On second thought, this will be an inconvenience to the Prove method, as it can preprocess the circuit.

We assume that the Transcript has been seeded.